lib/pbench/agent/tool_meister_client.py

```diff
@@ -87,149 +79,120 @@ def __init__(
-                    host=self.redis_host, port=self.redis_port, db=0
-                )
-            except Exception as e:
-                self.logger.error(
-                    "Unable to connect to redis server, %s:%s: %s",
-                    self.redis_host,
-                    self.redis_port,
-                    e,
-                )
-                raise
-            else:
-                self.logger.debug("constructed Redis() object")
-
-        if self.to_client_chan is None:
-            self.to_client_chan = RedisChannelSubscriber(
-                self.redis_server, self.to_client_channel
+        """On context entry, setup the connection with the Redis server using
+        a SignalExporter instance."""
+        if self.redis_server:
+            self.logger.debug(
+                "constructing SignalExporter() object using existing Redis"
+                " connection, name: %s",
+                self.publisher_name,
+            )
+            sig_pub = state_signals.SignalExporter(
+                self.publisher_name, existing_redis_conn=self.redis_server
+            )
+        else:
+            self.logger.debug(
+                "constructing SignalExporter() object using host %s:%s," " name: %s",
+                self.redis_host,
+                self.redis_port,
+                self.publisher_name,
+            )
+            sig_pub = state_signals.SignalExporter(
+                self.publisher_name,
+                redis_host=self.redis_host,
+                redis_port=self.redis_port,
             )
+        sig_pub.initialize_and_wait(
+            1,
+            list(tm_allowed_actions),
+            tag="from_pbench_client",
+        )
+        self.sig_pub = sig_pub
+        self.logger.debug("constructed SignalExporter() object")
         return self
 
     def __exit__(self, *args):
-        """On context exit, just close down the to client channel object."""
-        if self.to_client_chan is not None:
-            self.to_client_chan.close()
+        """On context exit, just close down the to SignalExporter object."""
+        self.sig_pub.shutdown()
 
-    def publish(self, group, directory, action, args=None):
-        """publish - marshal a JSON document formed from the group, directory,
-        action, and args arguments.
+    def publish(self, group: str, directory: str, action: str, args=None) -> int:
+        """Publish a state signal formed from the group, directory, action,
+        and args arguments, using the SignalExporter instance.  It waits for
+        responses from subscribers (TDS, and the TMs indirectly) before
+        continuing.
 
         Returns 0 on success, 1 on failure; logs are also written for any
         errors encountered.
+
         """
-        if action not in api_tm_allowed_actions:
+        if action not in tm_allowed_actions:
+            self.logger.warning(f"Attempted to publish illegal action '{action}'")
             return 1
```

Comment:
Should Pbench still be responsible for checking and enforcing this?  (Didn't we enable State Signals to do this...or what was the point of passing `tm_allowed_actions` to `SignalExporter.initialize_and_wait()` at line 116?)

Reply:
@Maxusmusti, do you know the answer to this?

Reply:
Yeah, this is a great point, the check has become redundant, as state-signals already handles it. The original logic was to keep it in at the beginning for an even faster failure, but upon further inspection it really looks like it's doing more harm than good.